File: lib/java/src/lib/Answer.java
package lib;

import lombok.experimental.UtilityClass;

@UtilityClass
public class Answer {

    public void timer(Runnable solution) {
        long start = System.nanoTime();
        solution.run();
        long end = System.nanoTime();
        System.out.printf("Runtime (ns): %d\n", end - start);
    }

    public <T> void part1(T expected, T actual) {
        part(1, expected, actual);
    }

    public <T> void part2(T expected, T actual) {
        part(2, expected, actual);
    }

    private <T> void part(int part, T expected, T actual) {
        if (!expected.equals(actual)) {
            var errorFormat = "Part %d: expected %s got %s";
            throw new RuntimeException(String.format(errorFormat, part, expected, actual));
        }
        System.out.printf("Part %d: %s\n", part, actual);
    }
}

File: lib/elixir/answer.ex
defmodule Answer do
  @spec part1!(t, t) :: :ok when t: any()
  def part1!(expected, actual), do: part!(1, expected, actual)

  @spec part2!(t, t) :: :ok when t: any()
  def part2!(expected, actual), do: part!(2, expected, actual)

  @spec part!(1 | 2, t, t) :: :ok when t: any()
  defp part!(n, expected, actual) do
    if expected == actual do
      IO.puts("Part #{n}: #{actual}")
    else
      raise "Part #{n}: expected #{expected} got #{actual}"
    end
  end
end
